### PR TITLE
[TensorFlow] Adding LeViT

### PR DESCRIPTION
**Pull request by ariG23498**
# Adding TensorFlow version of LeViT

This PR adds the TensorFlow version of [LeViT](https://arxiv.org/abs/2104.01136).

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. Issue linked: #19123 
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by ariG23498** [comment]
Still working on it.

File: src/transformers/models/levit/modeling_tf_levit.py [hunk 1]
```diff
@@ -0,0 +1,1048 @@
...
+        # The epsilon and momentum used here are the defaults in torch batch norm layer.
+        self.batch_norm = tf.keras.layers.BatchNormalization(epsilon=1e-05, momentum=0.9, name="batch_norm")
+
+    def call(self, embeddings: tf.Tensor, training: Optional[bool] = None):
+        # embeddings shape = (bsz, num_channels, height, width)
+        embeddings = tf.transpose(embeddings, perm=(0, 2, 3, 1))
+        embeddings = self.padding(embeddings)
+        embeddings = self.convolution(embeddings, training=training)
+        embeddings = self.batch_norm(embeddings, training=training)
+        # embeddings shape = (bsz, height, width, num_channels)
```
Comment (sayakpaul) [inline src/transformers/models/levit/modeling_tf_levit.py:124]: Won't it be a channels-first memory layout after the transpose?

File: src/transformers/models/levit/modeling_tf_levit.py [hunk 2]
```diff
@@ -0,0 +1,1048 @@
...
+            name="convolution",
+        )
+        # The epsilon and momentum used here are the defaults in torch batch norm layer.
+        self.batch_norm = tf.keras.layers.BatchNormalization(epsilon=1e-05, momentum=0.9, name="batch_norm")
+
+    def call(self, embeddings: tf.Tensor, training: Optional[bool] = None):
+        # embeddings shape = (bsz, num_channels, height, width)
+        embeddings = tf.transpose(embeddings, perm=(0, 2, 3, 1))
+        embeddings = self.padding(embeddings)
+        embeddings = self.convolution(embeddings, training=training)
```
Comment (sayakpaul) [inline src/transformers/models/levit/modeling_tf_levit.py:122]: I am assuming you've verified if this is indeed the sequence to follow when matching a `torch.nn.Conv2d()` with `padding` defined?

File: src/transformers/models/levit/modeling_tf_levit.py [hunk 3]
```diff
@@ -0,0 +1,1048 @@
...
+            kernel_size=kernel_size,
+            strides=stride,
+            dilation_rate=dilation,
+            groups=groups,
+            use_bias=False,
+            data_format="channels_last",
+            name="convolution",
+        )
+        # The epsilon and momentum used here are the defaults in torch batch norm layer.
+        self.batch_norm = tf.keras.layers.BatchNormalization(epsilon=1e-05, momentum=0.9, name="batch_norm")
```
Comment (sayakpaul) [inline src/transformers/models/levit/modeling_tf_levit.py:116]: As per the official docs (https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html), these are the defaults:

* `eps=1e-05, momentum=0.1`

File: src/transformers/models/levit/modeling_tf_levit.py [hunk 4]
```diff
@@ -0,0 +1,1048 @@
...
+        in_channels,
+        out_channels,
+        kernel_size,
+        stride,
+        padding,
+        dilation=1,
+        groups=1,
+        bn_weight_init=1,
+        *args,
+        **kwargs,
```
Comment (sayakpaul) [inline src/transformers/models/levit/modeling_tf_levit.py:100]: Is there a reason why to have separate args and then kwargs? Usually, we only take kwargs here.

**Review by sayakpaul — COMMENTED** [review]
@ariG23498 see if changing the BN defaults fixes your issue.

File: src/transformers/models/levit/modeling_tf_levit.py (hunk 1 above)
Comment (ariG23498) [reply src/transformers/models/levit/modeling_tf_levit.py:124]: The `embeddings` that come in are channel-first. I use the first `transpose` operation to make the embeddings channel-last. This helps with the padding, conv and batch_norm. After applying the operations I also use the next `transpose` to make the computeed embeddings channel-first again.

Am I missing something here?

File: src/transformers/models/levit/modeling_tf_levit.py (hunk 2 above)
Comment (ariG23498) [reply src/transformers/models/levit/modeling_tf_levit.py:122]: https://github.com/huggingface/transformers/blob/7032e0203262ebb2ebf55da8d2e01f873973e835/src/transformers/models/regnet/modeling_tf_regnet.py#L66

I have actually used this comment and colab notebook linked.

File: src/transformers/models/levit/modeling_tf_levit.py (hunk 3 above)
Comment (ariG23498) [reply src/transformers/models/levit/modeling_tf_levit.py:116]: https://github.com/huggingface/transformers/blob/7032e0203262ebb2ebf55da8d2e01f873973e835/src/transformers/models/resnet/modeling_tf_resnet.py#L62-L63

I have found out that in this repo all the BatchNorm layers had `eps=1e-5 and momentum=0.9`.

File: src/transformers/models/levit/modeling_tf_levit.py (hunk 4 above)
Comment (ariG23498) [reply src/transformers/models/levit/modeling_tf_levit.py:100]: No specific reason. I can change it if you want!

Comment (sayakpaul) [reply src/transformers/models/levit/modeling_tf_levit.py:100]: No, it's like deviating from the standard implementations without reasoning. So.

File: src/transformers/models/levit/modeling_tf_levit.py (hunk 3 above)
Comment (sayakpaul) [reply src/transformers/models/levit/modeling_tf_levit.py:116]: Also, true. The default momentum clearly isn't 0.9. @amyeroberts any inputs here?

File: src/transformers/models/levit/modeling_tf_levit.py (hunk 1 above)
Comment (sayakpaul) [reply src/transformers/models/levit/modeling_tf_levit.py:124]: Got it. I thought you were referring to the shape after the transpose.

File: src/transformers/models/levit/modeling_tf_levit.py (hunk 3 above)
Comment (sayakpaul) [reply src/transformers/models/levit/modeling_tf_levit.py:116]: But did changing the defaults help in the successful cross-loading by any chance?

Comment (ariG23498) [reply src/transformers/models/levit/modeling_tf_levit.py:116]: It did not help me with anything in specific. I just did what all of the BatchNorm layers used.

Comment (sayakpaul) [reply src/transformers/models/levit/modeling_tf_levit.py:116]: What's the current issue again? 

Would be better to have a detailed trace along with the steps to reproduce it somewhere.

File: src/transformers/models/levit/modeling_tf_levit.py (hunk 4 above)
Comment (ariG23498) [reply src/transformers/models/levit/modeling_tf_levit.py:100]:
```python
from transformers import LevitFeatureExtractor, LevitModel, TFLevitModel
import torch
import tensorflow as tf
import numpy as np
from datasets import load_dataset

dataset = load_dataset("huggingface/cats-image")
image = dataset["test"]["image"][0]

feature_extractor = LevitFeatureExtractor.from_pretrained("facebook/levit-128S")

# Models
model_tf = TFLevitModel.from_pretrained("facebook/levit-128S", from_pt=True)
model_pt = LevitModel.from_pretrained("facebook/levit-128S")

# Inputs
inputs_tf = feature_extractor(image, return_tensors="tf")
inputs_pt = feature_extractor(image, return_tensors="pt")

# Outputs
outputs_tf = model_tf(**inputs_tf, training=False, output_hidden_states=False)
with torch.no_grad():
    outputs_pt = model_pt(**inputs_pt, output_hidden_states=False)

# Assertion
check_tf = outputs_tf.last_hidden_state
check_pt = outputs_pt.last_hidden_state
np.testing.assert_allclose(
    check_tf,
    check_pt,
    rtol=1e-5,
    atol=1e-5,
)
```
This creates a 100% mismatch.

Comment (sayakpaul) [reply src/transformers/models/levit/modeling_tf_levit.py:100]: And there's no warning when you're doing `model_tf = TFLevitModel.from_pretrained("facebook/levit-128S", from_pt=True)`? Like certain weights weren't used and so on?

Comment (ariG23498) [reply src/transformers/models/levit/modeling_tf_levit.py:100]: All the weights are ported properly. The only thing that I see is with `batch_norm` parameters.

I have also followed [this comment](https://github.com/huggingface/transformers/pull/17554#issuecomment-1152191816) in order to see whether the parameters of the batch norm are important to be ported or not.

Comment (sayakpaul) [reply src/transformers/models/levit/modeling_tf_levit.py:100]: Hmm. If the weights have been cross-loaded successfully and the outputs are not being matched it means the intermediate computations are differing. 

In this case, I would try to pick a few of the modules from the PT implementation, which could lead to differences in the outputs in the respective TF implementation. 

Do you have any such modules in mind?

Comment (ariG23498) [reply src/transformers/models/levit/modeling_tf_levit.py:100]: The [TFLEvitResidulaLayer](https://github.com/ariG23498/transformers/blob/feat/levit-tf/src/transformers/models/levit/modeling_tf_levit.py#L491-L512) would be a good place to start with the inspection.

I am placing my bets on this module due to the use of random module.

Comment (sayakpaul) [reply src/transformers/models/levit/modeling_tf_levit.py:100]: If that's the case, I'd suggest starting your investigation from there. Specifically, ensure each intermediate operation in that module matches with one (PT) another (TF).

File: src/transformers/models/levit/modeling_tf_levit.py (hunk 3 above)
Comment (amyeroberts) [reply src/transformers/models/levit/modeling_tf_levit.py:116]: The momentum parameter (rather confusingly) is different for PyTorch and TensorFlow. For momentum `x` in PyTorch, the equivalent value in TensorFlow is `1 - x`.

Comment (sayakpaul) [reply src/transformers/models/levit/modeling_tf_levit.py:116]: My brain will remain dead for today.

**Comment by ariG23498** [comment]
Working!
